### DISCUSSION TITLE
I do not understand optionnal loading

### DISCUSSION THREAD

**Discussion by probakilla** [wbthomason/packer.nvim#962] [Q&A]
Hello !
I recently tried to play with optionnal plugin, but I did not managed to get it working (yet).
For example, I got a plugin (williamboman/nvim-lsp-installer), that i tried to load on specific file types, so I tried something like this : 
```
function load_plug_conf(mod_name)
    require('plugin_config.' .. mod_name)
end

...

use {
        "williamboman/nvim-lsp-installer",
        ft = source_code_files,
        config = load_plug_conf('lsp_installer')  -- In this file, I do a "require('nvim-lsp-installer')"
}
```

The problem is that when I start NeoVim, I got an error that say me that it cannot "require" the said package.
The "config" key is not supposed to be executed AFTER the plugin is loaded ? It looks like it is done before.

Furthermore, I also tried to play with the "opt" and "cmd" options, but I did understand how to get it working as well.
I read the README several times, but nothing helped me. Without any "ft", "opt" or "cmd" my configuration is working, but I wanted to load modules only when I need them.

Here is the repo of my full config : https://gitlab.com/J.pilleux/stow_dotfiles/-/tree/master/nvim/.config/nvim

Thanks in advance !

**Comment by arsham** [comment]
You should provide a function to the config property:
```lua
use({
        "williamboman/nvim-lsp-installer",
        ft = source_code_files,
        config = function() load_plug_conf('lsp_installer')  end,
})
```

**Reply by probakilla** [reply]
Thanks for the answer ! That is a start, but when I try to import some configuration from an other file, that does not work.
For example, I got a file that looks like this : 

```lua
local lspinstaller = require("lsp_installer")
lsp_installer_cfg = {
    -- Some config
}
lspinstaller.setup(cfg)
```

And now I tried to use it like this : 
```lua
        "williamboman/nvim-lsp-installer",
        ft = source_code_files,
        config = function() require(’main.lazyloading.lsp_installer')  end,  -- does not work
```

It looks like it does not evaluate the content of my file, or its too late to do it.
I tried to just paste the whole file content inside the "config function" and that worked.
I also tried to just "export" the dict part of the configuration and I was able to retreive it :
```lua
-- lsp_installer.lua
local M = {}
M.cfg = { 
-- Some config
}
return M

-- packer.lua
 config = function() 
    local cfg = require(’main.lazyloading.lsp_installer')  
    require("lsp_config").setup(cfg) -- that worked
end,
```

The problem is when I got more code than just the "setup" (like additional keymaps), I wanted to split my configuration into more files to keep it "easy" to read.

The last thing I tried is just put the whole file inside an "init" function, export it like I did with the configuration, and then call it in my packer.lua, but without success.

I am able to make what I want by just writing the full configuration inside the "config" function, but I wanted to know if there is a way to make a more "elegent" packer file by putting all complex configuration into separate files.

Again, my actual nvim configuration is here (can change frequently as I am trying new things) : https://gitlab.com/J.pilleux/stow_dotfiles/-/tree/master/nvim/.config/nvim

Thanks

**Reply by arsham** [reply]
I'm afraid I'm not sure what is happening in your setup. I have my [setup](https://github.com/arsham/shark) in different files, with all the lazy loading and setup/config loaded from different places.
